### PR TITLE
Add Support for FSDP

### PR DESCRIPTION
**Pull request by pglusacTT**
### Issue
N/A

### Problem description
Per-device DRAM is becoming a bottleneck for large-scale experiments.
We need to reduce DRAM usage in experiments using Torch-XLA FSDP sharding.

### What's Changed
- Add FSDP wrapper to device manager
- Add logic for resolving strategy

### Swiss report
Megatron:
image

FSDP:
image

## Note: 
There is an issue with loss values we're observing that seems unrelated to this change.
When we aren't getting infs, the loss curve looks fine and loss values are in range.
Adding this PR to enable other experiments to utilize FSDP.
Tracking issue: https://github.com/tenstorrent/tt-blacksmith/issues/542

W&B report:
image

### Checklist
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by mmilosevicTT — APPROVED** [review]
LGTM

File: blacksmith/tools/device_manager.py [hunk 1]
```diff
@@ -125,6 +134,22 @@ def _apply_tensor_parallelism(self, model: nn.Module) -> nn.Module:
         torch_xla.sync(wait=True)
         return model
 
+    def _apply_fsdp(self, model: nn.Module) -> nn.Module:
+        # TODO(pglusac): Add support for FSDP granularity configuration.
+        def shard_output(output, mesh):
+            # Extract logits and shard its batch dimension along the fsdp axis.
+            real_output = getattr(output, "logits", None)
+            if real_output is None:
+                real_output = output[0] if isinstance(output, tuple) else output
```
Comment (mmilosevicTT) [inline blacksmith/tools/device_manager.py:143]: Does it have to be a tuple? Can it ba a list?

File: blacksmith/tools/device_manager.py [hunk 2]
```diff
@@ -94,13 +97,19 @@ def is_tensor_parallel(self) -> bool:
...
+        return self.mesh is not None and "fsdp" in self.mesh.axis_names
+
     def shard_tensor(self, tensor: torch.Tensor, sharding_spec: Tuple):
         return xs.mark_sharding(tensor, self.mesh, sharding_spec)
 
     def shard_model(self, model: nn.Module) -> nn.Module:
         """Shard model based on mesh configuration."""
+        if self.is_fsdp():
+            model = self._apply_fsdp(model)
         if self.is_tensor_parallel():
```
Comment (mmilosevicTT) [inline blacksmith/tools/device_manager.py:111]: Are we sure fsdp will shard each layer? Should we do elif just in case?

**Review by mmilosevicTT — COMMENTED** [review]
Left two questions just to confirm.

File: blacksmith/experiments/torch/gpt_oss/lora/loudbox/fsdp_gpt_oss_20b_sst2.yaml [hunk 3]
```diff
@@ -0,0 +1,67 @@
+# DISCLAIMER: This produces -inf in the loss.
```
Comment (agobeljicTT) [inline blacksmith/experiments/torch/gpt_oss/lora/loudbox/fsdp_gpt_oss_20b_sst2.yaml:1]: is there a case where it wont inf?

File: blacksmith/models/torch/gpt_oss/model_overrides.py [hunk 4]
```diff
@@ -9,8 +9,12 @@
 from transformers.utils.quantization_config import Mxfp4Config
 
 
-def get_model(config, device):
-    """Load GPT-OSS model with deinterleaving overrides, LoRA, and compilation."""
+def get_model(config, device, device_manager=None):
```
Comment (agobeljicTT) [inline blacksmith/models/torch/gpt_oss/model_overrides.py]: can we just use device_manager.device? device be redundant here

File: blacksmith/models/torch/gpt_oss/model_overrides.py [hunk 5]
```diff
@@ -40,6 +44,9 @@ def get_model(config, device):
 
     model.to(device)
 
+    if device_manager is not None:
+        model = device_manager.shard_model(model)
```
Comment (agobeljicTT) [inline blacksmith/models/torch/gpt_oss/model_overrides.py:49]: do we need to remove shard_model from other experiments?

File: blacksmith/experiments/torch/gpt_oss/train.py [hunk 6]
```diff
@@ -180,7 +178,6 @@ def train(
                 }
                 # Shard batch if data parallelism is used.
                 batch = device_manager.prepare_batch(batch)
-                device_manager.shard_model(model)
```
Comment (agobeljicTT) [inline blacksmith/experiments/torch/gpt_oss/train.py:183]: so this was done on each cycle of training?

File: blacksmith/tools/device_manager.py (hunk 2 above)
Comment (pglusacTT) [reply blacksmith/tools/device_manager.py:111]: I deliberately put it this way. FSDP will do it's thing, shard the entire model but you still have a choice to override shardings.

File: blacksmith/tools/device_manager.py (hunk 1 above)
Comment (pglusacTT) [reply blacksmith/tools/device_manager.py:143]: Sure, will extend.

File: blacksmith/experiments/torch/gpt_oss/lora/loudbox/fsdp_gpt_oss_20b_sst2.yaml (hunk 3 above)
Comment (pglusacTT) [reply blacksmith/experiments/torch/gpt_oss/lora/loudbox/fsdp_gpt_oss_20b_sst2.yaml:1]: I put up the w&b logs. You can see that it infs for some steps, not for others.

File: blacksmith/models/torch/gpt_oss/model_overrides.py (hunk 5 above)
Comment (pglusacTT) [reply blacksmith/models/torch/gpt_oss/model_overrides.py:49]: I want this to be optional for now, might do it later.

File: blacksmith/experiments/torch/gpt_oss/train.py (hunk 6 above)
Comment (pglusacTT) [reply blacksmith/experiments/torch/gpt_oss/train.py:183]: Yes but it doesn't seem to be necessary.

File: blacksmith/models/torch/gpt_oss/model_overrides.py (hunk 5 above)
Comment (agobeljicTT) [reply blacksmith/models/torch/gpt_oss/model_overrides.py:49]: add a TODO or a comment so we know this is deliberate

File: blacksmith/experiments/torch/gpt_oss/train.py
Comment (umalesTT) [inline blacksmith/experiments/torch/gpt_oss/train.py:139]: Should we put this in yaml file?

Comment (umalesTT) [inline blacksmith/experiments/torch/gpt_oss/train.py]: nit: we don't have per-step logging in other experiments, i guess this got left from debugging

File: blacksmith/models/torch/huggingface/hf_models.py [hunk 7]
```diff
@@ -27,6 +27,9 @@ def get_model(config: TrainingConfig, device: torch.device):
     model.to(eval(config.dtype))
     model.to(device)
 
+    if device_manager is not None:
```
Comment (umalesTT) [inline blacksmith/models/torch/huggingface/hf_models.py]: Can we do `get_model(config, device_manager, some_flag)`?
If I'm not mistaken you use `device_manager` as that flag here to make this optional - if we come up with another name we won't have redundant `device` arg and a more clear name for the flag

File: blacksmith/tools/device_manager.py (hunk 2 above)
Comment (umalesTT) [reply blacksmith/tools/device_manager.py:111]: For my understanding, if we wanted every odd layer with FSDP, we would FSDP the whole model and override shardings of even layers?

File: blacksmith/tools/device_manager.py [hunk 8]
```diff
@@ -125,6 +134,22 @@ def _apply_tensor_parallelism(self, model: nn.Module) -> nn.Module:
...
+            real_output = getattr(output, "logits", None)
+            if real_output is None:
+                real_output = output[0] if isinstance(output, tuple) else output
+            # Skip if the sharding annotation is already set.
+            if torch_xla._XLAC._get_xla_sharding_spec(real_output) not in (None, ""):
+                return
+            partition_spec = ("fsdp",) + (None,) * (real_output.dim() - 1)
+            xs.mark_sharding(real_output, mesh, partition_spec)
+
+        model = FSDP(model, mesh=self.mesh, shard_output=shard_output)
```
Comment (umalesTT) [inline blacksmith/tools/device_manager.py:151]: Why do we shard output by default?
I feel like we should have an option to do FSDP only , irregardless of output sharding

**Review by umalesTT — APPROVED** [review]
🔥

File: blacksmith/experiments/torch/gpt_oss/train.py
Comment (pglusacTT) [reply blacksmith/experiments/torch/gpt_oss/train.py:139]: I don't think it should reside in the yaml file. This is something that is model specific and not related to the experiment itself.

File: blacksmith/models/torch/huggingface/hf_models.py (hunk 7 above)
Comment (pglusacTT) [reply blacksmith/models/torch/huggingface/hf_models.py]: Makes sense, will do it.

File: blacksmith/tools/device_manager.py (hunk 2 above)
Comment (pglusacTT) [reply blacksmith/tools/device_manager.py:111]: Yes, we could do that as well, didn't think of it!

File: blacksmith/tools/device_manager.py (hunk 8 above)
Comment (pglusacTT) [reply blacksmith/tools/device_manager.py:151]: Ahhh it is buggy when sharding output is not provided. I'll add a todo for it but yeah.
